### PR TITLE
Fix OS-specific code (Mac/UNIX overlap)

### PR DESCRIPTION
**Pull request by UniqMartin**
*Note that the code from this pull request shouldn't be merged just yet, it is only a quick fix and intended as a basis for further discussion.*

The problem is rather simple: With the addition of the `Q_OS_*` macros for Qt5-compatibility, there is now an overlap between `Q_OS_MAC` (defined on OS X) and `Q_OS_UNIX` (also defined on OS X). Since the code blocks guarded by these defines are usually not mutually exclusive, sometimes the code for both OS X and Linux/UNIX runs. This breaks the OS X build as, e.g., `TWUtils::getLibraryPath` now starts to return `~/.TeXworks` instead of `/Library/TeXworks`.

If I'm not mistaken, TeXworks has three major platforms it supports: Windows, OS X and “Other” (Linux, BSD, etc.). I'd suggest to create three macros `TW_OS_WIN`, `TW_OS_MAC`, and `TW_OS_OTHER`, that are mutually exclusive and are defined in a central header (which one would that be?) that is included by *every* other source code file. These macros would then be used everywhere instead of the `Q_WS_*` and `Q_OS_*` macros used right now. Would that be acceptable? If so, I'd be glad to do the work and adapt this pull request.

If the above solution isn't acceptable, maybe it would be best to drop all references to `Q_WS_*` (`Q_OS_*` already exists in all versions of Qt4 I could get a hold of) and make the existing checks a bit more specific. I'd volunteer to do this, if this is the preferred solution.

### CODE REVIEW

**Comment by stloeffler** [comment]
I agree that we should have some simple and universal macros.

Here are a few random thoughts, a conclusion can be found at the end:
 * I'm not too fond of having a central file that all other files include. I know something like a config.h is used in many projects, but a simple, unrelated change (e.g., formatting, whitespaces, whatever) needlessly triggers a complete rebuild of the whole project. We had something like this with TWUtils.h which was included in many files and depended on SvnRev.h, which in turn got updated on every commit. Also, including this file is bound to be forgotten in some file or other in the future. Instead, we could define the appropriate flags in CMake?
 * Q_OS_UNIX is defined on the Mac for a reason: Mac OS X is derived from Unix. Thus, if you want to test whether to expect a Windows or *nix filesystem, for example, this can actually be helpful.
 * the name TW_OS_OTHER seems too generic to me.

However, Qt already has that system in place with Q_OS_*. Qt 4.5 doesn't seem to have Q_OS_MAC (http://doc.qt.digia.com/4.5/functions.html), but I guess we can make Qt 4.6 (released in 2009) or later the requirement without causing too many problems (or even Qt 4.7, released in 2010). This could also simplify some QT_VERSION checks that are in the sources right now.

In conclusion, I guess the easiest option would be to drop Q_WS_* in most places. AFAICT, Q_WS_* is intended to test for a particular UI, like X11, and not for an OS. So whenever we want to test for an OS, we should use Q_OS_* anyway. Plus, of course, use #if defined(Q_OS_MAC) #elif defined(Q_OS_UNIX) || defined(Q_OS_LINUX) #elif defined(Q_OS_WIN32 #endif to properly handle the Q_OS_UNIX on Mac issue.

**Comment by UniqMartin** [comment]
Thanks for your detailed response! This is very helpful, since I'm rather new to Qt development. My conclusion from this and some more research was that it's probably best to drop `Q_WS_*` and rely solely on `Q_OS_*`. This simplifies the code and also fixes the bug, that prompted me to open this pull request in the first place.

But I'm also in support of raising the Qt4 requirement to at least 4.6. If you support this move, I will follow up with a new pull request that does just that and then simplifies/improves the code as much as possible.

*Commit message from the revised and rebased fix with some additional details:*

The fix is related to the fact that `Q_OS_UNIX` is also defined on OS X, so in order to properly check for Linux and other non-OS-X UNIXes one has to check for `Q_OS_UNIX` and absence of `Q_OS_DARWIN` (a synonym for `Q_OS_MAC`, but `Q_OS_MAC` is only defined for Qt >= 4.6).

The simplification is about eliminating double checks for both `Q_WS_*` and `Q_OS_*` (the latter were added for Qt5 compatibility). The idea is to reduce all checks to `Q_OS_*`, more precisely:

* OS X: `defined(Q_OS_DARWIN)`
* Windows: `defined(Q_OS_WIN)`
* Linux et al.: `defined(Q_OS_UNIX) && !defined(Q_OS_DARWIN)`

Additional notes:

* The previously used check for both `Q_OS_UNIX` and `Q_OS_LINUX` was redundant as `Q_OS_UNIX` is basically defined as `!Q_OS_WIN` (checked with Qt 4.8.6 and 5.4.1, minus unsupported platforms like MS-DOS and OS/2).

* The check for `Q_OS_DARWIN` can be replaced with `Q_OS_MAC` as soon as the Qt requirement is raised to >= 4.6. This can be done with some simple search and replace and should make the code easier to understand for future contributors.

**Comment by UniqMartin** [comment]
I hope I'm not being impatient. Is there anything I can do to help get this (and my other pull request) merged? Thanks!

**Comment by stloeffler** [comment]
Sorry for the delay. Last week I've mostly been working on the Windows compilation issue (http://tug.org/pipermail/texworks/2015q2/006265.html). Anyway, it's merged now :)

**Comment by UniqMartin** [comment]
No problem and thanks for merging! I was just curious about the current status and was really hoping to get this merged, because my local changes (that I hope to contribute soon) build on these fixes. Now things are easier to manage for me again. :)
